pkg/binding/binding.go

```diff
@@ -69,6 +69,8 @@ func (c *Controller) syncBinding(ctx context.Context, bindingName string) error
 		return fmt.Errorf("failed to get BindingPolicy from informer cache (name=%v): %w", bindingName, policyErr)
 	}
 
+	policyErrors := extraValidateBindingPolicy(policy)
```

Comment:
I missed the changes that follow this line (in previously merged PRs). Managing a BP in this call violates the design of objects being updated only if their keys are queued and picked by a worker.

I think all the parts that perform checks and updates on BPs should be moved to `Controller::SyncBindingPolicy` and instead this code should queue the key when necessary. Otherwise, looks good.

Reply:
Yes, I definitely agree that some refactoring is in order. I am not so sure that it will be easy until we fully address a larger issue: making BindingPolicy just one way to manage a Binding. In any case, yes, for later PRs.

Reply:
Would be good to track in an issue.

Reply:
#2502